tests/integration/long/upgrade/__init__.py

```diff
@@ -0,0 +1,179 @@
+    will clean the environment and start the appropriate C* version according
+    to the upgrade path. The upgrade can be done in a different thread using the
+    start_upgrade upgrade_method (this would be the most realistic scenario)
+    or node by node, waiting for the upgrade to happen, using _upgrade_one_node method
+    """
+    UPGRADE_PATH = None
+    start_cluster = True
+
+    @classmethod
+    def setUpClass(cls):
+        cls.logger_handler = MockLoggingHandler()
+        logger = logging.getLogger(cluster.__name__)
+        logger.addHandler(cls.logger_handler)
+
+    def _setUp(self):
+        """
+        This is not the regular _setUp method because it will be called from
+        the decorator instead of letting nose handle it.
+        This setup method will start a cluster with the right version according
+        to the variable UPGRADE_PATH.
+        """
+        remove_cluster()
+        self.cluster = use_cluster(UPGRADE_CLUSTER + self.UPGRADE_PATH.name, [3],
+                                   ccm_options=self.UPGRADE_PATH.starting_version,
+                                   configuration_options=self.UPGRADE_PATH.configuration_options)
+        self.nodes = self.cluster.nodelist()
+        self.last_node_upgraded = None
+        self.upgrade_done = Event()
+        self.upgrade_thread = None
+
+        if self.start_cluster:
+            setup_keyspace()
+
+            self.cluster_driver = Cluster()
+            self.session = self.cluster_driver.connect()
+            self.logger_handler.reset()
+
+    def _tearDown(self):
+        """
+        special tearDown method called by the decorator after the method has ended
+        """
+        if self.upgrade_thread:
+            self.upgrade_thread.join(timeout=5)
+            self.upgrade_thread = None
+
+        if self.start_cluster:
+            self.cluster_driver.shutdown()
+
+    def start_upgrade(self, time_node_upgrade):
+        """
+        Starts the upgrade in a different thread
+        """
+        self.upgrade_thread = Thread(target=self._upgrade, args=(time_node_upgrade,))
+        self.upgrade_thread.start()
+
+    def _upgrade(self, time_node_upgrade):
+        """
+        Starts the upgrade in the same thread
+        """
+        start_time = time.time()
+        while self._upgrade_one_node():
+            end_time = time.time()
+            time_to_upgrade = end_time - start_time
+            if time_node_upgrade > time_to_upgrade:
+                time.sleep(time_node_upgrade - time_to_upgrade)
+        self.upgrade_done.set()
+
+    def is_upgraded(self):
+        """
+        Returns True if the upgrade has finished and False otherwise
+        """
+        return self.upgrade_done.is_set()
+
+    def wait_for_upgrade(self, timeout=None):
+        """
+        Waits until the upgrade has completed
+        """
+        self.upgrade_done.wait(timeout=timeout)
+
+    def _upgrade_one_node(self):
```

Comment:
I don't love how much state this needs to maintain -- maybe a better design would be `def upgrade_node(self, node_id): ...` and callers can use

```python
for node_id in self.nodes:
    self.upgrade_node(node_id)
    ...
```

Reply:
Hmm, yeah I did it this way because I figured it wouldn't be important which node is upgraded, you just want to get the next node upgraded. To me it's a trade-off of adding more complexity to be more flexible.

Reply:
Yeah, I guess we have different views of complexity -- but you're right that this is the only method that uses this state, so it's not too hard to read. :+1: